### PR TITLE
:sparkles: allow un-reconciling (unlocking) transactions

### PR DESCRIPTION
**Pull request by MatissJanis**
In the past month I've found myself duplicating a locked transaction just so I could amend some details. Which is a bit annoying.. 

Hence my recommendation would be to allow unlocking transactions by clicking on the "lock" icon.

cc @youngcw @zachwhelchel LMK what you think

### CODE REVIEW

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 1]
```diff
@@ -1349,7 +1354,14 @@ const Transaction = memo(function Transaction(props) {
   );
 });
 
-function TransactionError({ error, isDeposit, onAddSplit, onDistributeRemainder, style, canDistributeRemainder }) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:1352]: All of these other changes in the file: automatically applied by prettier. Just code formatting changes.

I wonder why linter was not breaking about them.. weird

**Comment by youngcw** [comment]
I would prefer it to be a bit harder to unlock the transaction.  It takes some work to lock, so it should also take some work to unlock imho.

I think it would work well to add an option in the transaction edit menu along with the bulk edit options

**Comment by MatissJanis** [comment]
@youngcw I'm not quite sure I understand the rationale for explicitly adding user friction. Can you explain your thought process: why would adding more friction here be beneficial? Especially for actions that are undo-able by ctrl+z.

**Comment by zachwhelchel** [comment]
@MatissJanis why do you need to duplicate a transaction to make changes? You can still make changes to a locked transaction it just pops up a confirm message. I'd think that's easier than duplicating? Am I missing something about your use case?

Screenshot 2024-01-19 at 3 42 28 PM

**Comment by MatissJanis** [comment]
@zachwhelchel date cannot be changed on reconciled transactions (which makes sense). Also automated imports won't update reconciled transactions (which again: totally makes sense).

Is there a good reason why we would prohibit folks from unlocking? I've read a couple of support threads on Discord where the recommendation was to "just duplicate the transaction". Which doesn't scream to me like particularly friendly solution.

**Comment by Teprifer** [comment]
My two cents, locking is about preventing unintentional or casual changes, such as imports, bulk actions(could add a prompt/notice if the selected transactions contain one or more that are locked, more sure on current state), and drawing a line between a 'known good' point in time, and what's in flux.

But this is still user data they control, so I don't see why they shouldn't be to easily unlock it to make an intentional change.

**Comment by MatissJanis** [comment]
cc also @shall0pass I think you had some strong opinions about this too.

--

Either way, even if I sound very direct here in my opinion - I don't have any objections to closing off this PR without merging and leaving the status-quo as is. Lets have a conversation and decide :) 

--

Related:
https://github.com/actualbudget/actual/issues/2110

**Comment by zachwhelchel** [comment]
It's also worth nothing that we were kind of talking past each other because there is a bug. When I try to change the date on a locked transaction (maybe it's Safari?) it let's me with a warning. When @MatissJanis does the same it just doesn't do anything. So to him it felt like a ton of friction to have to "duplicate" it just to make that change. To me it didn't feel that heavy because mine is working without the bug.

**Comment by youngcw** [comment]
> @youngcw I'm not quite sure I understand the rationale for explicitly adding user friction. Can you explain your thought process: why would adding more friction here be beneficial? Especially for actions that are undo-able by ctrl+z.

I feel like if its too easy to undo a reconcile then it makes the reconcile lose its usefulness.  I feel like the reconcile is about creating a psuedo fresh start and if I can unlock things too easy it makes my new foundation not mean anything.

A compromise might be to have one of the edit warnings for unlocking instead of hiding the option in a menu.  Im not against being able to unlock, it is the users data after all, I just think it needs to be appropriately deliberate of an action.

**Comment by youngcw** [comment]
@zachwhelchel fixing that bug does sound like a good place to start before making other changes

**Comment by shall0pass** [comment]
I've come around on this a bit.  I'm having a hard time justifying the difference between being able to unlock the transaction manually vs deleting the transaction and recreating it.  I think if the warning were presented during an "unlock" action, that would be sufficient as "unlocking" a transaction could bring your reconciliation out of balance in the same way as deleting a transaction would.

However, I do think that the duplicate feature we have shouldn't be duplicating a locked or cleared status.  Duplicating a reconciled transaction where it copies over as locked also brings the reconciliation out of balance.

That's my 2 cents.  Overall, I'm in favor of the change to allow unlocking of a reconciled transaction.

**Comment by youngcw** [comment]
> Duplicating a reconciled transaction where it copies over as locked also brings the reconciliation out of balance.

I agree this needs fixed.  There is an issue for it #2155. @zachwhelchel would you be willing to to look at it, and maybe the date issue?

**Comment by MatissJanis** [comment]
Changes:
- added a warning dialog when un-locking transactions

Screenshot 2024-01-21 at 15 10 03

**Comment by shall0pass** [comment]
I just have one more thought.  Once this dialog is clicked through and the reconciliation is out of balance, I'm not sure it's entirely necessary to display it again until after the account is re-reconciled.  Thanks for working on this Matias.

**Comment by MatissJanis** [comment]
> I just have one more thought. Once this dialog is clicked through and the reconciliation is out of balance, I'm not sure it's entirely necessary to display it again until after the account is re-reconciled. Thanks for working on this Matiss.

That should already be the case in this PR, no? After unlocking the warning is no longer shown if you modify notes/payee/amount/other field.

**Comment by shall0pass** [comment]
> That should already be the case in this PR, no? After unlocking the warning is no longer shown if you modify notes/payee/amount/other field.

I was thinking more globally. If you edit one transaction and "void" the reconciliation, is there a purpose to the dialog on a different locked transaction.  I'm not thinking that kind of change would be in this PR. Just a thought that occured to me.

**Comment by youngcw** [comment]
Im fine with this state.  I think trying to track a global reconciled state would just cause confusion when the warnings all of a sudden don't show anymore.

**Review by youngcw — APPROVED** [review]

**Pull request merged**